### ISSUE TITLE
Incorrect reportGeneralType "unknown symbol import" errors when type checking is enabled

### ISSUE THREAD

**Issue by onet-git** [microsoft/pylance-release#471]
## Environment data

-   Language Server version: 2010.10.0
-   OS and version: Ubuntu Xenial
-   Python version (& distribution if applicable, e.g. Anaconda): 3.5

## Expected behaviour

```python
from production import foo
from production import generated
```

Files `production/foo.py` and `bazel-bin/production/generated.py` exist.

If `"python.analysis.extraPaths": []` then pylance should report "unknown symbol import" on `generated`.

If `"python.analysis.extraPaths": [ "bazel-bin"]` then pylance should not report an error.

## Actual behaviour

If `"python.analysis.extraPaths": []` then pylance reports "unknown symbol import" on `generated`

If `"python.analysis.extraPaths": [ "bazel-bin"]` then pylance reports "unknown symbol import" on `foo`

If `"python.analysis.extraPaths": [ ".", "bazel-bin"]` then pylance reports "unknown symbol import" on `generated`

If `"python.analysis.extraPaths": [ "bazel-bin", "."]` then pylance reports "unknown symbol import" on `foo`

## Code Snippet / Additional information

I thought this was a PEP420 issue, but it when type checkings disabled the problem goes away. Maybe type checking doesn't implement PEP420 correctly.

**Comment by erictraut** [comment]
Thanks for reporting the issue.

When typeCheckingMode is set to "off", many diagnostic rules are disabled. The error is still detected, but it's suppressed. That explains why the issue apparently goes away.

A couple of follow-up questions for you.

Am I correct in assuming that these are namespace packages — i.e. there is not an `__init__.py` file within the top-level `production` directory?

Also, your problem description talks about the symbol `bar`, but this isn't mentioned in the sample you provided. Can you explain where `bar` is defined and how it's imported or referenced?

**Comment by erictraut** [comment]
Also, it looks like your extraPaths don't include the correct path to the `bazel-bin` directory. Try the following and see if it works.
`"python.analysis.extraPaths": ["production/bazel-bin/"]`
Note that you don't need to specify "." because that path is already assumed.

**Comment by onet-git** [comment]
I made some typos in the original bug reports, this example is a namspace package.

Attached is a workspace that demonstrates the issue. It includes a log file which clearly shows the example case where having `python.analysis.extraPaths` set results in pylance not even looking for `production/foo.py`.

[pylance-bug-471.zip](https://github.com/microsoft/pylance-release/files/5356327/pylance-bug-471.zip)

**Comment by erictraut** [comment]
Thanks for providing the example. Very helpful. I was able to repro the bug and find a fix. 

The issue is that pylance aggressively caches any package import resolutions. This is a big performance boost, so it's important to do. In general, it's safe to cache imported packages because the Python interpreter also does so. However, there's a catch. With namespace packages, the Python interpreter isn't able to cache the results. Pylance was still doing aggressive caching for namespace package resolutions. After the first successful import from "production", all subsequent attempts to import from "production" resolved to the same namespace package even if it couldn't provide the imported symbol. That explains the behavior you were seeing.

I've made the caching mechanism in Pylance smarter so it can detect this situation and deal with it appropriately. This fix will be in the next release.
